src/lib/update-lock.ts

```diff
@@ -82,6 +82,75 @@ async function dispose(
 	}
 }
 
+/**
+ * Given a lockfile path `p`, return a tuple [appId, serviceName] of that path.
+ * Paths are assumed to end in the format /:appId/:serviceName/(resin-)updates.lock.
+ */
+function getIdentifiersFromPath(p: string) {
+	const parts = p.split('/');
+	if (parts.pop()?.match(/updates\.lock/) === null) {
+		return [];
+	}
+	const serviceName = parts.pop();
+	const appId = parts.pop();
+	return [appId, serviceName];
+}
+
+type LockedEntity = { appId: number; services: string[] };
+
+/**
+ * A map of locked services by appId.
+ * Exported for tests only; getServicesLockedByAppId is the public generator interface.
+ */
+export class LocksTakenMap extends Map<number, Set<string>> {
+	constructor(lockedEntities: LockedEntity[] = []) {
+		// Construct a Map<number, Set<string>> from user-friendly input args
+		super(
+			lockedEntities.map(({ appId, services }) => [appId, new Set(services)]),
+		);
+	}
+
+	// Add one or more locked services to an appId
+	public add(appId: number, services: string | string[]): void {
```

Comment:
Since you are extending `Map` I think you can override `set` here

Reply:
I didn't do that as the functionality of `add` behaves different than `Map.prototype.set`: `add` adds to the list of locks that are taken for an appId, whereas `Map.prototype.set` replaces the value for a key. I'd prefer to keep it this way for that reason, but open to discussion of course.